UefiTestingPkg/Library/FlatPageTableLib/AArch64/FlatPageTableAArch64.c

```diff
@@ -0,0 +1,371 @@
+    EntryCount = TT_ENTRY_COUNT;
+  }
+
+  for (Index = 0; Index < EntryCount; Index++, RegionStart += RegionLength) {
+    // Skip unmapped entries
+    if (!IS_VALID (PagingEntry[Index].Uint64)) {
+      continue;
+    }
+
+    if (IS_BLOCK (PagingEntry[Index].Uint64, Level)) {
+      // FUTURE WORK: Should we check the feature register to see if blocks are allowed at level 1?
+      ASSERT (Level == 1 || Level == 2 || Level == 3);
+
+      ScratchEntry.Uint64 = PagingEntry[Index].Tteb.Uint64;
+
+      // If the entry is a block, then then some access attributes are inherited from the parent
+      // when FEAT_HPDS is active
+      if (mHierarchicalControlEnabled) {
+        // Check if the parent table entry has XnTable bits to pass down to the block entry.
+        //
+        // This logic was derived from the Armv8 A Architecture Manual version H.a
+        // Section D5.4.5: Data access permission controls
+        // Subsection: Hierarchical control of instruction fetching
+        //
+        // If in EL2, only the XN bit is valid
+        if ((ArmReadCurrentEL () == AARCH64_EL2) && (ParentHeritableAttributes.Bits.XnTable != 0)) {
+          ScratchEntry.Tteb.Bits.Xn = 1;
+        }
+        // if in EL1, both UXN and PXN bits are valid
+        else if ((ArmReadCurrentEL () == AARCH64_EL1)) {
+          // Xn is Uxn for EL1
+          if (ParentHeritableAttributes.Bits.XnTable != 0) {
+            ScratchEntry.Tteb.Bits.Xn = 1;
+          }
+
+          if (ParentHeritableAttributes.Bits.PxnTable != 0) {
+            ScratchEntry.Tteb.Bits.Pxn = 1;
+          }
+        }
+
+        // Check if the parent table entry has ApTable bits to pass down to the block entry
+        //
+        // This logic was derived from the Armv8 A Architecture Manual version H.a
+        // Section D5.4.5: Data access permission controls
+        // Subsection: Hierarchical control of data access permissions
+        //
+        // 0b01 -> Access from EL0 is not allowed, no effect on write permissions
+        if ((ParentHeritableAttributes.Bits.ApTable & TT_TABLE_AP_MASK) == TT_TABLE_AP_EL0_NO_ACCESS) {
+          // BIT6 toggles access from EL0
+          ScratchEntry.Tteb.Bits.AccessPermissions &= ~((UINT64)BIT6); // Clear BIT6
+          // 0b10 -> Access from EL0 is read-only, no write permissions at any EL
+        } else if ((ParentHeritableAttributes.Bits.ApTable & TT_TABLE_AP_MASK) == TT_TABLE_AP_NO_WRITE_ACCESS) {
+          // BIT7 toggles write access for all ELs
+          ScratchEntry.Tteb.Bits.AccessPermissions |= BIT7; // Set BIT7
+          // 0b11 -> Access from EL0 is not allowed, no write permissions at any EL
+        } else if ((ParentHeritableAttributes.Bits.ApTable & TT_TABLE_AP_MASK) == TT_TABLE_AP_MASK) {
+          ScratchEntry.Tteb.Bits.AccessPermissions |= BIT7;            // Set BIT7
+          ScratchEntry.Tteb.Bits.AccessPermissions &= ~((UINT64)BIT6); // Clear BIT6
+        }
+      }
+
+      if ((*LastEntry != NULL) &&

```

Comment:
LastEntry is not NULL checked before it is dereferenced here.

Reply:
Added a NULL check on LastEntry at the top of the function